substrafl/strategies/fed_pca.py

```diff
@@ -0,0 +1,107 @@
+from substrafl.nodes.train_data_node import TrainDataNode
+from substrafl.remote import remote
+from substrafl.schemas import FedAvgAveragedState
+from substrafl.schemas import FedAvgSharedState
+from substrafl.schemas import StrategyName
+from substrafl.strategies import FedAvg
+
+
+class FedPCA(FedAvg):
+    """Federated Principal Component Analysis strategy."""
+
+    def __init__(self):
+        super(FedPCA, self).__init__()
+
+        # current local and share states references of the client
+        self._local_states: Optional[List[LocalStateRef]] = None
+        self._shared_states: Optional[List[SharedStateRef]] = None
+
+    @property
+    def name(self) -> StrategyName:
+        """The name of the strategy
+
+        Returns:
+            StrategyName: Name of the strategy
+        """
+        return StrategyName.FEDERATED_PCA
+
+    def perform_round(
+        self,
+        algo: Algo,
+        train_data_nodes: List[TrainDataNode],
+        aggregation_node: AggregationNode,
+        round_idx: int,
+        clean_models: bool,
+        additional_orgs_permissions: Optional[set] = None,
+    ):
+        """One round of the Federated PCA"""
+        if aggregation_node is None:
+            raise ValueError("In FedAvg strategy aggregation node cannot be None")
+
+        if round_idx == 0:
+            # Initialization of the strategy by performing a local update on each train data organization
+            assert self._local_states is None
+            assert self._shared_states is None
+            self._perform_local_updates(
+                algo=algo,
+                train_data_nodes=train_data_nodes,
+                current_aggregation=None,
+                round_idx=round_idx,
+                aggregation_id=aggregation_node.organization_id,
+                additional_orgs_permissions=additional_orgs_permissions or set(),
+                clean_models=clean_models,
+            )
+            function_to_execute = self.avg_shared_states
+        else:
+            function_to_execute = self.avg_shared_states_with_qr
+
+        current_aggregation = aggregation_node.update_states(
+            function_to_execute(shared_states=self._shared_states, _algo_name="Aggregating"),  # type: ignore
+            round_idx=round_idx,
+            authorized_ids={train_data_node.organization_id for train_data_node in train_data_nodes},
+            clean_models=clean_models,
+        )
+
+        self._perform_local_updates(
+            algo=algo,
+            train_data_nodes=train_data_nodes,
+            current_aggregation=current_aggregation,
+            round_idx=round_idx,
+            aggregation_id=aggregation_node.organization_id,
+            additional_orgs_permissions=additional_orgs_permissions or set(),
+            clean_models=clean_models,
+        )
+
+    @remote
+    def avg_shared_states_with_qr(self, shared_states: List[FedAvgSharedState]) -> FedAvgAveragedState:
```

Comment:
Adding a test on this function to check that if computes the shared states the right way would be great (create a test_fed_pca.py file in test/strategies)

Reply:
+1